### PR TITLE
Add --deps flag

### PR DESCRIPTION
**Pull request by mirkojotic**
Initial work on getting --deps flag to print module dependencies.

I was just wondering if I'm heading in the right direction ( and I could always use some advice ).

I'm positive that my messing around in `getModuleDependencies` wasn't right so any pointers around that would be greatly appreciated.

No tests so far :( ( also I would appreciate some advice as to where to place them )

Solves #630 ( well it doesn't yet :D )

### CODE REVIEW

File: js/compiler.ts [hunk 1]
```diff
@@ -400,7 +401,8 @@ export class DenoCompiler implements ts.LanguageServiceHost {
    */
   getModuleDependencies(
     moduleSpecifier: ModuleSpecifier,
-    containingFile: ContainingFile
+    containingFile: ContainingFile,
+    byModuleName = false
```
Comment (kitsonk) [inline js/compiler.ts]: I don't think we need a non-by module name variant of this...  The only reason it was ever here was to support the module names, which we didn't have exposed... I would just always assume it will be by module name.

File: js/compiler.ts [hunk 2]
```diff
@@ -84,7 +84,8 @@ export class ModuleMetaData implements ts.IScriptSnapshot {
   constructor(
     public readonly fileName: ModuleFileName,
     public readonly sourceCode = "",
-    public outputCode = ""
+    public outputCode = "",
+    public moduleName = ""
```
Comment (kitsonk) [inline js/compiler.ts]: I would change the order... `moduleName` shouldn't be optional and should be either first or second, I think it should also be `public readonly`.

Also, would `moduleId` be better here.  That is more commonly what it is referred to as.  It should also be of type `ModuleId` which should be a type alias for `string`.  So it would mean the following:

* `fileName` - An absolute resource identifier, which is often the local cache file name.
* `moduleId` - The absolute module ID, which could be a URL or a local filename.
* `moduleSpecifier` - A potentially relative module name, which is resolved in relationship to...
* `containingFile` - The containing module of which the module specifier is resolved against.

**Comment by mirkojotic** [comment]
@kitsonk thank you very much for all the feedback. I'll wait for ry to chime in and then continue working on it!

Side question. For --deps-json would I just recursively call the same method for all found dependencies?

**Comment by ry** [comment]
@mirkojotic Do you know `.d` dependency files - as outputted by compilers and used by tools like make and ninja? It shows the dependency structure of `#include`. Rust also can output these. 

For example, here is `out/debug/rust_crates/tokio.d`
```make
rust_crates/libtokio.rlib: ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/lib.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/clock.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/scheduler.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/fs.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/net.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/reactor/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/reactor/poll_evented.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/builder.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/builder.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/shutdown.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/task_executor.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/timer.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/util/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/util/future.rs

rust_crates/tokio.d: ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/lib.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/clock.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/scheduler.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/fs.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/net.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/reactor/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/reactor/poll_evented.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/builder.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/builder.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/shutdown.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/task_executor.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/timer.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/util/mod.rs ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/util/future.rs

../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/lib.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/clock.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/mod.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/scheduler.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/fs.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/net.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/reactor/mod.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/reactor/poll_evented.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/mod.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/builder.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/mod.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/builder.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/shutdown.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/task_executor.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/timer.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/util/mod.rs:
../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/util/future.rs:
```

I think it would be useful for deno to output this:
1. It's a cross-tool way of specifying the dep structure.
2. It's minimal in its syntax

Ideally http module specifiers would be absolute paths to .deno/deps, not literal URLs.

What do you think?

**Comment by mirkojotic** [comment]
Hey @ry I did not but I do now :) 

Yeah, this looks like an interesting solution and I can see the benefits, I'd love to work on it. Couple of questions:
1. In _addition to_ or _instead of_ `--deps` flag that will only log out dependencies in a list?
2. Whould `--depfile` actually run the module after outputting dependencies or would it just write the file and exit?

**Comment by ry** [comment]
> In addition to or instead of --deps flag that will only log out dependencies in a list?

@mirkojotic In addition. Let's land this one first and do --depfile as a separate PR.

> Whould --depfile actually run the module after outputting dependencies or would it just write the file and exit?

I think we should support both. But let's go with just writing it first.

**Comment by mirkojotic** [comment]
@kitsonk I've made a couple of changes based on your feedback. When you get a minute could you review it and let me know what else needs work. After you're satisfied we can ping ry for final review.

File: src/flags.rs [hunk 3]
```diff
@@ -33,8 +34,9 @@ pub fn print_usage() {
 -v or --version    Print the version.
 -r or --reload     Reload cached remote resources.
 -D or --log-debug  Log debug output.
--h or --help       Print this message.
---v8-options       Print V8 command line options."
+-h or --help             Print this message.
```
Comment (ry) [inline src/flags.rs]: Bad indent here

File: js/compiler.ts [hunk 4]
```diff
@@ -489,6 +495,7 @@ export class DenoCompiler implements ts.LanguageServiceHost {
     if (fileName && this._moduleMetaDataMap.has(fileName)) {
       return this._moduleMetaDataMap.get(fileName)!;
     }
+    let moduleId: ModuleId = "";
     let sourceCode: string | undefined;
```
[the remark this answers is not in the document]
Comment (ry) [reply js/compiler.ts]: Maybe just another commit in this PR, if you're comfortable with rebase and squash

[the remark this answers is not in the document]
Comment (mirkojotic) [reply js/compiler.ts]: Yup! I sure will!

File: js/compiler_test.ts [hunk 5]
```diff
@@ -430,7 +430,7 @@ test(function compilerGetModuleDependencies() {
     "foo/baz.ts",
     "/root/project"
   );
-  assertEqual(deps, ["/root/project/foo/bar.ts", "/root/project/foo/baz.ts"]);
+  assertEqual(deps, ["foo/bar", "foo/baz"]);
```
[the remark this answers is not in the document]
Comment (mirkojotic) [reply js/compiler_test.ts]: That's what I was thinking but for whatever reason I only get relative file specifier back and without extension... When I run it from the command line it works as expected. Any advice?

[the remark this answers is not in the document]
Comment (kitsonk) [reply js/compiler_test.ts]: @mirkojotic yes, you need to update the first argument of the fixtures here:

https://github.com/denoland/deno/blob/10dc71133af211de40e29e7a7ab8c470a4a2c417/js/compiler_test.ts#L110-L142

That mocks out the return from `os.codeFetch`.

**Comment by mirkojotic** [comment]
git mishap.. :( will be resolve soon.

File: js/global-eval.ts [hunk 6]
```diff
@@ -1,6 +1,10 @@
-// If you use the eval function indirectly, by invoking it via a reference
-// other than eval, as of ECMAScript 5 it works in the global scope rather than
-// the local scope. This means, for instance, that function declarations create
-// global functions, and that the code being evaluated doesn't have access to
-// local variables within the scope where it's being called.
+/**
```
Comment (kitsonk) [inline js/global-eval.ts]: Hmmm... this shouldn't be part of this PR... seems like something strange with the rebasing again.

File: js/util.ts [hunk 7]
```diff
@@ -3,53 +3,67 @@ import { TypedArray } from "./types";
 
 let logDebug = false;
 
+// @internal
```
Comment (kitsonk) [inline js/util.ts]: Again, shouldn't be part of this PR.

**Comment by mirkojotic** [comment]
Hey @kitsonk sorry about the commit last night. I had some issues rebasing :( I'm managed to squash everything. **Should** be ok now.

**Comment by mirkojotic** [comment]
@kitsonk @ry thanks for being patient. I'll start working on --depfile next.
